browser/search_engines/search_engine_provider_service.h

```diff
@@ -8,43 +9,42 @@
 #include <memory>
 #include <string>
 
-#include "base/macros.h"
+#include "base/scoped_observation.h"
 #include "components/keyed_service/core/keyed_service.h"
 #include "components/prefs/pref_member.h"
+#include "components/search_engines/template_url_service.h"
+#include "components/search_engines/template_url_service_observer.h"
 
 class Profile;
 class TemplateURL;
 class TemplateURLService;
 
-class SearchEngineProviderService : public KeyedService {
+// Handles DDG toggle button in NTP of guest window.
+class SearchEngineProviderService : public KeyedService,
+                                    public TemplateURLServiceObserver {
  public:
-  explicit SearchEngineProviderService(Profile* otr_profile);
+  explicit SearchEngineProviderService(Profile* profile);
   ~SearchEngineProviderService() override;
+  SearchEngineProviderService(const SearchEngineProviderService&) = delete;
+  SearchEngineProviderService& operator=(const SearchEngineProviderService&) =
+      delete;
 
- protected:
-  // If subclass want to know and configure according to prefs change, override
-  // this.
-  virtual void OnUseAlternativeSearchEngineProviderChanged() {}
-
-  bool UseAlternativeSearchEngineProvider() const;
-  void ChangeToAlternativeSearchEngineProvider();
-  void ChangeToNormalWindowSearchEngineProvider();
-
-  // Points off the record profile.
-  Profile* otr_profile_;
-  // Service for original profile of |otr_profile_|.
-  TemplateURLService* original_template_url_service_;
-  // Service for off the record profile.
-  TemplateURLService* otr_template_url_service_;
-
-  std::unique_ptr<TemplateURL> alternative_search_engine_url_;
+  // TemplateURLServiceObserver overrides:
+  void OnTemplateURLServiceChanged() override;
 
  private:
-  void OnPreferenceChanged(const std::string& pref_name);
-
+  Profile* profile_;
+  TemplateURLService* template_url_service_;
+  std::unique_ptr<TemplateURL> alternative_search_engine_url_;
   BooleanPrefMember use_alternative_search_engine_provider_;
+  base::ScopedObservation<TemplateURLService, TemplateURLServiceObserver>
+      observation_{this};
```

Comment:
Is this how you are able to track when person switches between a private window and a regular window? (for the text that shows in the omnibox)

Reply:
This `SearchEngineProviderService` class is only for handling guest profile window's search provider.
Guest profile window is handled differently with other profile (normal, private(tor)) because guest profile is completely different profile with normal profile.
Other profile window's search provider is handled by `ActiveWindowSearchProviderManager`.

Reply:
Ah.. forgot to handling omnibox text. :)

Reply:
Fixed by https://github.com/brave/brave-core/pull/8539/commits/6691c448e13c5650025f512231eaa135194118ec